src/persist-client/src/stats.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Aggregate statistics about data stored in persist.
+
+use std::collections::BTreeMap;
+
+use mz_persist::indexed::columnar::ColumnarRecords;
+use mz_persist_types::columnar::{ColumnFormat, PartEncoder, Schema};
+use mz_persist_types::part::{Part, PartBuilder};
+use mz_persist_types::Codec;
+
+use crate::internal::encoding::Schemas;
+
+/// Logic for filtering parts in a read as uninteresting purely from stats about
+/// the data held in them.
+pub trait PartFilter<K, V>: 'static {
+    /// True if the part should be fetched, false if it should be skipped.
+    fn should_fetch(&mut self, stats: &PartStats) -> bool;
+}
+
+/// An implementation of [PartFilter] that always returns true.
+#[derive(Debug)]
+pub struct FetchAllPartFilter;
+
+impl<K, V> PartFilter<K, V> for FetchAllPartFilter {
+    fn should_fetch(&mut self, _stats: &PartStats) -> bool {
+        true
+    }
+}
+
+/// Aggregate statistics about data contained in a [Part].
+///
+/// TODO(mfp): PartialEq, Eq, PartialOrd, Ord are sus.
+#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct PartStats {
+    // TODO(mfp): Only contains Option<u64> columns in this initial skeleton.
+    pub(crate) key_col_min_max_nulls: BTreeMap<String, (u64, u64, usize)>,
+}
+
+impl PartStats {
+    pub(crate) fn new<K: Codec, V: Codec>(
+        schemas: &Schemas<K, V>,
+        part: &Part,
+    ) -> Result<Self, String> {
+        // TODO(mfp): This impl is all entirely just a placeholder. Zero chance
+        // it works like this by the time it ships.
+        let mut key_cols = part.key_ref();
+        let mut key_col_min_max_nulls = BTreeMap::new();
+        for (name, typ) in schemas.key.columns() {
+            let col = match (typ.optional, typ.format) {
+                (true, ColumnFormat::U64) => key_cols.col::<Option<u64>>(name.as_str())?,
+                // TODO(mfp): Support stats on other column types.
+                _ => continue,
+            };
+            let (mut min, mut max, mut nulls) = (u64::MAX, 0u64, 0usize);
+            for x in col.iter() {
+                if let Some(x) = x {
+                    min = std::cmp::min(*x, min);
+                    max = std::cmp::max(*x, max);
+                } else {
+                    nulls += 1;
+                }
+            }
+            key_col_min_max_nulls.insert(name, (min, max, nulls));
```

Comment:
wondering if we want min and max to be optional separate from null count. if the column is only nulls, it seems odd to report a min of u64::MAX and a max of 0

(I realize I'm making this comment on a bit of code marks `This impl is all entirely just a placeholder. Zero chance it works like this by the time it ships.` 😁)

Reply:
yeah, my inclination is to not bikeshed the specifics of the placeholder impl, but also maybe I'm wrong and this sort of thing is exactly why we have code review

Reply:
oh! I could just skip it if min > max, that's easy enough